crates/subspace-networking/src/protocols/autonat_wrapper.rs

```diff
@@ -0,0 +1,150 @@
+        Self {
+            inner: Autonat::new(config.local_peer_id, config.inner_config.clone()),
+            config,
+            listen_addresses: Default::default(),
+        }
+    }
+
+    fn private_ips_enabled(&self) -> bool {
+        !self.config.inner_config.only_global_ips
+    }
+
+    fn address_corresponds_to_listening_addresses(&self, addr: &Multiaddr) -> bool {
+        let candidate_protocol = addr
+            .iter()
+            .find_map(|protocol| match protocol {
+                udp @ Protocol::Udp(_) => Some(udp),
+                tcp @ Protocol::Tcp(_) => Some(tcp),
+                _ => None,
+            })
+            .expect("Either TCP or UDP protocol should be enabled.");
+
+        let address_result = self
+            .listen_addresses
+            .iter()
+            .any(|addr| addr.iter().any(|protocol| protocol == candidate_protocol));
+
+        debug!(
+            %address_result,
+            ?addr,
+            listen_addresses=?self.listen_addresses,
+            "Address candidate corresponds to listening addresses."
+        );
+
+        address_result
+    }
+
+    pub(crate) fn public_address(&self) -> Option<&Multiaddr> {
+        self.inner.public_address()
+    }
+
+    pub(crate) fn confidence(&self) -> usize {
+        self.inner.confidence()
+    }
+}
+
+impl NetworkBehaviour for Behaviour {
+    type ConnectionHandler = <Autonat as NetworkBehaviour>::ConnectionHandler;
+    type ToSwarm = AutonatEvent;
+
+    fn handle_established_inbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        local_addr: &Multiaddr,
+        remote_addr: &Multiaddr,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.inner.handle_established_inbound_connection(
+            connection_id,
+            peer,
+            local_addr,
+            remote_addr,
+        )
+    }
+
+    fn handle_established_outbound_connection(
+        &mut self,
+        connection_id: ConnectionId,
+        peer: PeerId,
+        addr: &Multiaddr,
+        role_override: Endpoint,
+    ) -> Result<THandler<Self>, ConnectionDenied> {
+        self.inner
+            .handle_established_outbound_connection(connection_id, peer, addr, role_override)
+    }
+
+    fn on_swarm_event(&mut self, event: FromSwarm) {
+        match event {
+            new_listen_addr_event @ FromSwarm::NewListenAddr(_) => {
+                if let FromSwarm::NewListenAddr(addr) = new_listen_addr_event {
+                    //TODO: handle listener address change
+                    self.listen_addresses.insert(addr.addr.clone());
+
+                    if self.private_ips_enabled() || is_global_address_or_dns(addr.addr) {
+                        self.inner.on_swarm_event(new_listen_addr_event);
+                    } else {
+                        debug!(addr=?addr.addr, "Skipped listening address in AutonatWrapper.");
+                    }
+                }
+            }
```

Comment:
This is not necessary to process in a special way, just forward it down to autonat